pkg/filter/filter_test.go

```diff
@@ -0,0 +1,349 @@
+			Expect(err).NotTo(HaveOccurred())
+
+			uo := unstructured.Unstructured{Object: mo}
+
+			err = filter.SetLabel(&uo)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(uo.GetLabels()).To(HaveKeyWithValue("specialresource.openshift.io/owned", "true"))
+		})
+	})
+
+	Context("IsSpecialResource", func() {
+		cases := []struct {
+			name    string
+			obj     client.Object
+			matcher types.GomegaMatcher
+		}{
+			{
+				name: "SpecialResource",
+				obj: &v1beta1.SpecialResource{
+					TypeMeta: metav1.TypeMeta{Kind: "SpecialResource"},
+				},
+				matcher: BeTrue(),
+			},
+			{
+				name: "Pod owned by SRO",
+				obj: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"specialresource.openshift.io/owned": "true"},
+					},
+				},
+				matcher: BeFalse(),
+			},
+			{
+				name: "valid selflink",
+				obj: func() *unstructured.Unstructured {
+					uo := &unstructured.Unstructured{}
+					uo.SetSelfLink("/apis/sro.openshift.io/v1")
+
+					return uo
+				}(),
+				matcher: BeTrue(),
+			},
+			{
+				name: "selflink in Label",
+				obj: func() *unstructured.Unstructured {
+					uo := &unstructured.Unstructured{}
+					uo.SetLabels(map[string]string{"some-label": "/apis/sro.openshift.io/v1"})
+
+					return uo
+				}(),
+				matcher: BeTrue(),
+			},
+			{
+				name:    "no selflink",
+				obj:     &unstructured.Unstructured{},
+				matcher: BeFalse(),
+			},
+		}
+
+		entries := make([]TableEntry, 0, len(cases))
+
+		for _, c := range cases {
+			entries = append(entries, Entry(c.name, c.obj, c.matcher))
+		}
+
+		DescribeTable(
+			"should return the correct value",
+			func(obj client.Object, m types.GomegaMatcher) {
+				Expect(filter.IsSpecialResource(obj)).To(m)
+			},
+			entries...)
+	})
+
+	Context("Owned", func() {
+		cases := []struct {
+			name    string
+			obj     client.Object
+			matcher types.GomegaMatcher
+		}{
+			{
+				name: "via ownerReferences",
+				obj: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						OwnerReferences: []metav1.OwnerReference{
+							{Kind: "SpecialResource"},
+						},
+					},
+				},
+				matcher: BeTrue(),
+			},
+			{
+				name: "via labels",
+				obj: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"specialresource.openshift.io/owned": "whatever"},
+					},
+				},
+				matcher: BeTrue(),
+			},
+			{
+				name:    "not owned",
+				obj:     &corev1.Pod{},
+				matcher: BeFalse(),
+			},
+		}
+
+		entries := make([]TableEntry, 0, len(cases))
+
+		for _, c := range cases {
+			entries = append(entries, Entry(c.name, c.obj, c.matcher))
+		}
+
+		DescribeTable(
+			"should return the expected value",
+			func(obj client.Object, m types.GomegaMatcher) {
+				Expect(filter.Owned(obj)).To(m)
+			},
+			entries...,
+		)
+	})
+
+	Context("Predicate", func() {
+		resetMode := func() { filter.Mode = "" }
+
+		BeforeEach(resetMode)
+		AfterEach(resetMode)
+
+		Context("CreateFunc", func() {
```

Comment:
If we said that `Context` is for the function level, e.g. 1 Context for each function, then why do we have nested contexts here?